# Tests for the NewsData API

In [25]:
import requests
import os

api_key = os.environ["CONTENT_CURATION_NEWSDATA_API_KEY"]

response = requests.get(f"https://newsdata.io/api/1/latest?apikey={api_key}&language=en&category=politics&prioritydomain=top&size=10")

In [26]:
import json
content = json.loads(response.content)

In [27]:
content

{'status': 'success',
 'totalResults': 2429,
 'results': [{'article_id': '448a2955c904938422fc8503ad540416',
   'title': 'FBI to announce new information in Portland, Vancouver ballot box arson cases',
   'link': 'https://www.kgw.com/article/news/politics/elections/fbi-new-information-portland-vancouver-ballot-box-arson/283-c8e1cec2-6410-4a63-8630-b5ded494aebd',
   'keywords': ['elections', 'local', 'politics', 'crime', 'news'],
   'creator': None,
   'video_url': None,
   'description': 'Incendiary devices were used to start fires inside three ballot boxes in the weeks before the 2024 election — two in Vancouver and one in Portland.',
   'content': 'ONLY AVAILABLE IN PAID PLANS',
   'pubDate': '2025-01-16 19:49:29',
   'pubDateTZ': 'UTC',
   'image_url': 'https://media.tegna-media.com/assets/KGW/images/b187fd5c-de02-4eaa-9e12-569aed7612a0/20241113T212641/b187fd5c-de02-4eaa-9e12-569aed7612a0_750x422.jpg',
   'source_id': 'kgw',
   'source_priority': 3758,
   'source_name': 'Kgw',
   's

In [34]:
import time
response = requests.get(f"https://newsdata.io/api/1/latest?apikey={api_key}&language=en&category=politics&prioritydomain=top&size=10")
content = json.loads(response.content)
pubdates = []

while len(content["results"]) > 0:
    for item in content["results"]:
        pubdates.append(item["pubDate"] + " " + item["pubDateTZ"])
    response = requests.get(f"https://newsdata.io/api/1/latest?apikey={api_key}&language=en&category=politics&prioritydomain=top&size=10&page={content['nextPage']}")
    content = json.loads(response.content)
    time.sleep(1)

TypeError: string indices must be integers

In [35]:
content

{'status': 'error',
 'results': {'message': 'Rate limit exceeded. Please try your request again later. For further details, please refer to our documentation at: https://newsdata.io/documentation/#rate-limit',
  'code': 'RateLimitExceeded'}}

In [33]:
pubdates

['2025-01-16 20:01:09',
 '2025-01-16 20:00:50',
 '2025-01-16 19:59:46',
 '2025-01-16 19:59:07',
 '2025-01-16 19:55:11',
 '2025-01-16 19:53:51',
 '2025-01-16 19:53:01',
 '2025-01-16 19:51:55',
 '2025-01-16 19:49:29',
 '2025-01-16 19:48:30',
 '2025-01-16 19:48:27',
 '2025-01-16 19:48:24',
 '2025-01-16 19:48:20',
 '2025-01-16 19:45:01',
 '2025-01-16 19:45:00',
 '2025-01-16 19:45:00',
 '2025-01-16 19:44:53',
 '2025-01-16 19:44:46',
 '2025-01-16 19:43:39',
 '2025-01-16 19:43:00',
 '2025-01-16 19:41:37',
 '2025-01-16 19:39:04',
 '2025-01-16 19:38:39',
 '2025-01-16 19:38:00',
 '2025-01-16 19:36:59',
 '2025-01-16 19:35:01',
 '2025-01-16 19:34:00',
 '2025-01-16 19:33:13',
 '2025-01-16 19:33:13',
 '2025-01-16 19:30:20',
 '2025-01-16 19:30:06',
 '2025-01-16 19:30:01',
 '2025-01-16 19:30:00',
 '2025-01-16 19:30:00',
 '2025-01-16 19:29:36',
 '2025-01-16 19:28:30',
 '2025-01-16 19:28:30',
 '2025-01-16 19:28:30',
 '2025-01-16 19:28:30',
 '2025-01-16 19:28:30',
 '2025-01-16 19:27:45',
 '2025-01-16 19:

Using the /latest API, news from the last 48 hours can be retrieved, but also delayed by 12 hours. 

It appears publish dates are sorted by most-recent-first, which makes working with the API a bit easier. 

In the above test, the API indicates 2429 results available. However, there is a limit of 200 requests per day meaning only 2000 of these results can actually be processed on any day. This is not an issue because the internal database will store some news stories, so there is no reason to read all the results given that the database already contains the earlier ones.

The actual content of the news story is not available: just the title, image, and description. The other APIs that I was evaluating also did not have the content available.